### DISCUSSION TITLE
Install 'next' to run from the command line?

### DISCUSSION THREAD

**Discussion by banshee** [vercel/next.js#16098] [Help]
What's the right way to install the 'next' command so you can run from the command line without npx?

I tried 

    npm -g install next react

But I get:

> james@gig:~/workspace/nexttutorial/nextjs-blog$ next dev
> The module 'react-dom' was not found. Next.js requires that you include it in 'dependencies' of your 'package.json'. To add it, run 'npm install react-dom'
> (node:104613) UnhandledPromiseRejectionWarning: Error: Cannot find module 'react-dom/server'
> Require stack:
> - /home/linuxbrew/.linuxbrew/lib/node_modules/next/dist/next-server/server/render.js
> - /home/linuxbrew/.linuxbrew/lib/node_modules/next/dist/next-server/server/next-server.js
> - /home/linuxbrew/.linuxbrew/lib/node_modules/next/dist/server/next.js
> - /home/linuxbrew/.linuxbrew/lib/node_modules/next/dist/server/lib/start-server.js
> - /home/linuxbrew/.linuxbrew/lib/node_modules/next/dist/cli/next-dev.js
> - /home/linuxbrew/.linuxbrew/lib/node_modules/next/dist/bin/next
>     at Function.Module._resolveFilename (internal/modules/cjs/loader.js:623:15)
>     at Function.Module._load (internal/modules/cjs/loader.js:527:27)
>     at Module.require (internal/modules/cjs/loader.js:681:19)
>     at require (internal/modules/cjs/helpers.js:16:16)
>     at Object.<anonymous> (/home/linuxbrew/.linuxbrew/lib/node_modules/next/dist/next-server/server/render.js:1:136)
>     at Module._compile (internal/modules/cjs/loader.js:774:30)
>     at Object.Module._extensions..js (internal/modules/cjs/loader.js:785:10)
>     at Module.load (internal/modules/cjs/loader.js:641:32)
>     at Function.Module._load (internal/modules/cjs/loader.js:556:12)
>     at Module.require (internal/modules/cjs/loader.js:681:19)
> (node:104613) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 1)
> (node:104613) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
> 

This works fine in that directory though:

> npx next dev
> ready - started server on http://localhost:3000
> event - compiled successfully
>

**Comment by lfades** [comment]
You need to have `next`, `react` and `react-dom` in the dependencies of the `package.json` in your app (more on that [here](https://nextjs.org/docs/getting-started#manual-setup)), then you can add scripts that use the [Next.js CLI](https://nextjs.org/docs/api-reference/cli). The CLI is not meant to be used globally outside of an app and you **should not** install `next` globally

**Reply by banshee** [reply]
I've read that page multiple times and it never explains how you get next on your path.  It jumps from using npx straight to using next, but there's no explanation of what to do to get next working.

> james@gig:~/workspace/nexttutorial/nextjs-blog$ npx next -h
> 
>     Usage
>       $ next <command>
> 
>     Available commands
>       build, start, export, dev, telemetry
> 
>     Options
>       --version, -v   Version number
>       --help, -h      Displays this message
> 
>     For more information run a command with the --help flag
>       $ next build --help
>   
> james@gig:~/workspace/nexttutorial/nextjs-blog$ next
> 
> Command 'next' not found, but can be installed with:
> 
> sudo apt install mailutils-mh  # version 1:3.7-2.1, or
> sudo apt install mmh           # version 0.4-2
> sudo apt install nmh           # version 1.7.1-6
> 
> james@gig:~/workspace/nexttutorial/nextjs-blog$ 
>

**Reply by lfades** [reply]
> I've read that page multiple times and it never explains how you get next on your path

You shouldn't be adding `next` to your path, that's why we don't mention it. Next.js is always tied to a single application, having it installed globally is highly unrecommended.

In other words, you should install a global module if you’re installing something that you want to use in your shell / command line (Next.js is not this, but a package like `create-next-app` that's used to initialize Next.js projects from the command line does apply here.)

And you should install a module locally if you’re installing something that you want to use in your **program / app**. This is the case of the `next` package.

**Comment by banshee** [comment]
Do people just do something like

```export PATH=./node_modules/.bin:$PATH```

in their .bashrc?

**Reply by lfades** [reply]
No this is not recommended, only packages that are intended to be used globally should be in your path

**Comment by banshee** [comment]
When I see this in the documentation https://nextjs.org/docs/advanced-features/debugging:

https://photos.app.goo.gl/1ygxvDwLgbAKVUD99
> Next.js being a Node.js application, all we have to do is to pass down the --inspect flag to the underlying Node.js process for it to start in debug mode.
> 
> First, start Next.js with the inspect flag:
> 
> NODE_OPTIONS='--inspect' next dev
>
> If you're using npm run dev or yarn dev (See: Getting Started) then you should update the dev script on your package.json:
> 
> 

I read that doc as being really clear about running ```next``` by itself from the command line, as opposed to using npx next or yarn next.  What's that doc supposed to be telling me?  Am I supposed to be reading that ```next``` command as a generic placeholder for ```npx next ...```?

**Reply by lfades** [reply]
Good point, this page is not clear about that. You should use the script as it's shown in the second code block:

```json
"dev": "NODE_OPTIONS='--inspect' next dev"
```

> That code should be inside `scripts` in your `package.json`

I'll make sure to update the docs according to your feedback. Let me know if there's anything else that's confusing about this.

**Reply by lfades** [reply]
Same goes for our CLI docs, I'll update those docs to be more clear about this. Also `npx next dev` and `npx next start` should indeed work like we show in the docs, what happens is that in order for it to work you need to run the command inside a project that has Next.js installed (which we don't mention right now).

**Reply by banshee** [reply]
> what happens is that in order for it to work you need to run the command inside a project that has Next.js installed (which we don't mention right now).

As a new user, that idea - you need to be in a project where next.js has been installed - I think is reasonably clear.  You might add something specific about not installing next.js globally.

You might think about something like changing to use ```$next``` (with the dollar sign) in the examples, to make it obvious that you don't mean that ```next``` is a command that can be run by itself.

**Reply by huperniketes** [reply]
This is still an issue.
